code/network/lib/Loot/Network/ZMQ/Client.hs

```diff
@@ -0,0 +1,270 @@
+
+    , ztPeers           :: TVar (Set ZTNodeId)
+      -- ^ List of peers we are connected to. Is to be updated by the
+      -- main thread only (broker worker).
+
+    , ztClients         :: TVar (Map ClientId ZTClientEnv)
+      -- ^ Channels binding broker to clients, map from client name to.
+    , ztSubscriptions   :: TVar (Map Subscription (Set ClientId))
+      -- ^ Map from subscription key to clents' identifiers.
+    , ztMsgTypes        :: TVar (Map MsgType ClientId)
+      -- ^ Map from msg type key to clents' identifiers.
+
+    , ztCliRequestQueue :: CliRequestQueue
+      -- ^ Queue to read (internal, administrative) client requests
+      -- from, like updating peers or resetting connection.
+    }
+
+instance HasLens ZTNetCliEnv r ZTNetCliEnv =>
+         HasLens CliRequestQueue r CliRequestQueue where
+    lensOf =
+        (lensOf @ZTNetCliEnv) .
+        (lens ztCliRequestQueue (\ztce rq2 -> ztce {ztCliRequestQueue = rq2}))
+
+createNetCliEnv :: MonadIO m => ZTGlobalEnv -> Set ZTNodeId -> m ZTNetCliEnv
+createNetCliEnv (ZTGlobalEnv ctx) peers = liftIO $ do
+    -- I guess it's alright to connect ROUTER instead of binding it.
+    -- https://stackoverflow.com/questions/16109139/zmq-when-to-use-zmq-bind-or-zmq-connect
+    ztCliBack <- Z.socket ctx Z.Router
+    forM_ peers $ Z.connect ztCliBack . ztNodeIdRouter
+
+    ztCliSub <- Z.socket ctx Z.Sub
+
+    ztClients <- newTVarIO mempty
+    ztPeers <- newTVarIO peers
+    ztSubscriptions <- newTVarIO mempty
+    ztMsgTypes <- newTVarIO mempty
+    ztCliRequestQueue <- CliRequestQueue <$> TQ.newTQueueIO
+
+    pure ZTNetCliEnv {..}
+
+----------------------------------------------------------------------------
+-- Methods
+----------------------------------------------------------------------------
+
+data CliBrokerStmRes
+    = CBClient ClientId (Maybe ZTNodeId) (MsgType, Content)
+    | CBBack
+    | CBSub
+    | CBRequest InternalRequest
+
+runBroker :: (MonadReader r m, HasLens' r ZTNetCliEnv, MonadIO m, MonadMask m) => m ()
+runBroker = do
+    ZTNetCliEnv{..} <- view $ lensOf @ZTNetCliEnv
+    -- This function may do something creative (LRU! Lowest ping!),
+    -- but instead (for now) it'll just choose a random peer.
+    let choosePeer = do
+            -- Yes, modulo bias. It shouldn't really matter.
+            i <- abs <$> randomIO
+            atomically $ do
+                l <- Set.toList <$> readTVar ztPeers
+                pure $ l L.!! (i `mod` length l)
+
+    let resolvePeer :: MonadIO m => ByteString -> m (Maybe ZTNodeId)
+        resolvePeer nodeid = do
+            atomically $ find ((== nodeid) . ztNodeConnectionId) <$> readTVar ztPeers
+
+    let sendToClient clientId (nId :: ZTNodeId, content :: CliRecvMsg) = do
+            res <- atomically $ do
+                cMap <- readTVar ztClients
+                let toWrite = bReceiveQ <$> Map.lookup clientId cMap
+                maybe pass (\tq -> TQ.writeTQueue tq (nId,content)) toWrite
```

Comment:
Please, refrain from coding like that. It is almost unreadable. Remember, that someone will have to debug that one day.